fastparquet/api.py

```diff
@@ -391,10 +392,94 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
         except IOError:
             pass
         self._set_attrs()
-
         if write_fmd:
             self._write_common_metadata(open_with)
 
+    def append_as_row_groups(self, data, row_group_offsets=None,
+                             compression=None, write_fmd:bool=True,
+                             open_with=default_open, mkdirs=None, append=True,
+                             stats=True):
+        """
+        Append data as new row groups to disk. Updated `ParquetFile` metadata
```

Comment:
Top of docstring should be kept to one line

Reply:
Reminds me: how is this method different from `write(..., append=True)` ?

Reply:
The differences with `write` are:
- it is used on an already instanciated `ParquetFile`, (if the instance is already existing/loaded, you don't re-load it again as opposed to `write`)
- there are fewer checks on the `ParquetFile` instance / it requires less parameters, as some of them are defined directly from the instance attributes (`partition_on` for instance)
- it leaves the option to write or not common metadata to disk (`write_fmd=False`)

After this PR, I would like to prepare a new one containing documentation and an illustrative test case. 
This will hopefully illustrate better the interest of completing the set of utilities to manage row groups of a `ParquetFile` instance.

This documentation would be a kind of 'tutorial' / example about how using these functions together to 
update a parquet dataset.

Assuming:
- the user defines a way to identify _overlaps_ in row groups.
- the user defines a way to identify _duplicates_ in rows.

Then the following _update methodology_ with fastparquet is possible:
```python
# Retrieve overlaps between existing row group and new data.
rgs_overlaps = filter_row_groups()

# Concat overlapping row groups with new data, and drop duplicates in rows.
df_to_add = pd.concat(new_data, pf[rgs_overlaps].to_pandas()).drop_duplicates()

# Remove overlaps from disk.
pf.remove_row_groups(rgs_overlaps, write_fmd=False)

# Add new data, along with data from overlapping row groups.
pf.append_as_row_groups(df_to_add)

# Sort row groups if relevant.
pf.fmd.row_groups = sorted(pf.row_groups, key=...)

# Finally write common metadata.
pf._write_common_metadata()
```
I would like to provide 2 application test cases:
- one with overlaps defined by partition. For instance:
`filter_row_groups(pd, filter=[('name', 'in', ['Yoh','Fred'])`
- the other one with overlaps defined by an ordered column. for instance a time series:
`filter_row_groups(pd, filter=[('timestamp', '>=', pd.Timestamp('2021/01/01'), ('timestamp', '<=', pd.Timestamp('2021/01/04')])`
and then derive the proposed procedure.

Compared to the `overwrite`, this approach:
- pros:
  - does not force the user to rely on partitions.
  - does not the limit the user to have a single row group per partitions
- cons:
  - requires more knowledge about these different row group utilities (hence I think a documentation page is worthwhile)
  - (similar) more lines of codes


I introduced this in #676, I am sorry if I am writing a lot of things not necessarily very clear.

Reply:
(made some updates to complete above answer)

Reply:
Hi @martindurant 
Replying here to
_By the way, is "insert" possible (i.e., shuffling the other row-groups and filenames as necessary)?_

I rephrased this question in my mind into:
_Could your work be extended so as to make insertion of row groups ('shuffling' other row-groups and filenames as necessary?_

To this question, my answer is yes, and I think it is actually a good idea!
It has not been my target so far (as it is possible to do it by using `sorted` on row group list directly), but I think doing insertion directly in `append_as_row_groups` (that I will rename into `write_row_groups` as per our above discussion) is a neater solution.

I think i would also change the interface of `write_row_groups()` so that it accepts an iterable of dataframes (e.g. list of dataframes or iterator of dataframes, each defining a row group) instead of a single dataframe to be split. for each of these row groups, the list of indexes where these row groups have to be inserted will have to be provided optionally (if not provided, it will be usual append).

With these changes, we can make the existing `overwrite` feature a separate function, similar in the way it is 'external' as the existing `merge`, and as efficient as it is now (but which implementation is intricated into `write`).

As I see it, it will thus bring us more modular code, easier to read and maintain, so yes, I am keen on investing this!
I will work out something in the coming days, and push updates, probably by the end of this week.
Thanks for your constructive insights!
Bests,